In [73]:
%reload_ext sql

In [ ]:
%sql mysql://admin:sql_2021@lmu-dev-01.cniirpmw0saj.us-east-1.rds.amazonaws.com/sql_project

# Primary Question: 
What Costco store in the Los Angeles area is best?

# Related Questions: 
What location in Los Angeles has the highest rating and supplies the largest selection of products? What location has the most employees working per rating?

# Exploratory Queries

# #1 
Purpose: This query can be helpful because it provides insight into employee compensation if an employee wants to understand their payment in relation to others as well as age, on average.

In [14]:
%%sql
SELECT 
	CONCAT(first_name, ' ', last_name) AS employee_full_name,
	age,
	AVG(salary) OVER(
        ORDER BY age
    )AS average_salary_per_age,
	CASE 
		WHEN age < (
			SELECT AVG(age)
			FROM employees e
		) THEN 'younger than average'
		WHEN age BETWEEN (
			SELECT AVG(age)
			FROM employees e
		) AND 50 THEN 'little over average age'
		ELSE 'older than average'
	END AS employee_age_dispersion
FROM employees e 
GROUP BY employee_full_name
ORDER BY salary DESC;

 * mysql://admin:***@lmu-dev-01.cniirpmw0saj.us-east-1.rds.amazonaws.com/sql_project
96 rows affected.


[('Charlie Parker', 30, Decimal('101384.5714'), 'younger than average'),
 ('Ellia Turner', 35, Decimal('90311.6364'), 'younger than average'),
 ('Maria Taylor', 30, Decimal('101384.5714'), 'younger than average'),
 ('Owen Howard', 36, Decimal('93367.3654'), 'younger than average'),
 ('Jordan Hawkins', 36, Decimal('93367.3654'), 'younger than average'),
 ('Dainton Payne', 37, Decimal('96121.3929'), 'younger than average'),
 ('Emily Stewart', 31, Decimal('100982.0000'), 'younger than average'),
 ('Martin Russell', 37, Decimal('96121.3929'), 'younger than average'),
 ('Miranda Ellis', 33, Decimal('91029.3667'), 'younger than average'),
 ('Natalie Moore', 40, Decimal('93416.8028'), 'little over average age'),
 ('Justin Sullivan', 36, Decimal('93367.3654'), 'younger than average'),
 ('Daisy Johnson', 35, Decimal('90311.6364'), 'younger than average'),
 ('Anna Spencer', 39, Decimal('93658.6212'), 'little over average age'),
 ('Miller Jones', 35, Decimal('90311.6364'), 'younger than average'),
 ('Andrew Craig', 39, Decimal('93658.6212'), 'little over average age'),
 ('Cherry Morris', 33, Decimal('91029.3667'), 'younger than average'),
 ('Abigail Harris', 33, Decimal('91029.3667'), 'younger than average'),
 ('Marcus Crawford', 36, Decimal('93367.3654'), 'younger than average'),
 ('Jaquelyn Rios', 32, Decimal('89467.0526'), 'younger than average'),
 ('Paul Alexander', 33, Decimal('91029.3667'), 'younger than average'),
 ('Darrel Callahan', 43, Decimal('93376.3378'), 'little over average age'),
 ('Harper Whitfield', 50, Decimal('91676.0854'), 'little over average age'),
 ('Christopher Barber', 52, Decimal('90848.5765'), 'older than average'),
 ('Brad Warren', 31, Decimal('100982.0000'), 'younger than average'),
 ('David Walker', 35, Decimal('90311.6364'), 'younger than average'),
 ('Hailey Myers', 37, Decimal('96121.3929'), 'younger than average'),
 ('Daquan Love', 30, Decimal('101384.5714'), 'younger than average'),
 ('Carlos Evans', 33, Decimal('91029.3667'), 'younger than average'),
 ('Reid Chapman', 34, Decimal('86063.8919'), 'younger than average'),
 ('Rylee Alston', 40, Decimal('93416.8028'), 'little over average age'),
 ('Derek Mitchell', 33, Decimal('91029.3667'), 'younger than average'),
 ('Ivana Vasquez', 46, Decimal('91594.1875'), 'little over average age'),
 ('Adrianna Sullivan', 40, Decimal('93416.8028'), 'little over average age'),
 ('Denise Coffey', 43, Decimal('93376.3378'), 'little over average age'),
 ('Sienna Miller', 34, Decimal('86063.8919'), 'younger than average'),
 ('Megan Baker', 45, Decimal('91689.0641'), 'little over average age'),
 ('Ted Hamilton', 32, Decimal('89467.0526'), 'younger than average'),
 ('Galvin Miles', 39, Decimal('93658.6212'), 'little over average age'),
 ('Joelle Mccray', 54, Decimal('90336.2955'), 'older than average'),
 ('Dana Haynes', 53, Decimal('90377.6322'), 'older than average'),
 ('Hu Odonnell', 34, Decimal('86063.8919'), 'younger than average'),
 ('Connie Phillips', 32, Decimal('89467.0526'), 'younger than average'),
 ('Dean Owen', 46, Decimal('91594.1875'), 'little over average age'),
 ('Melissa Armstrong', 32, Decimal('89467.0526'), 'younger than average'),
 ('Adam Stevens', 33, Decimal('91029.3667'), 'younger than average'),
 ('Ellia Moore', 37, Decimal('96121.3929'), 'younger than average'),
 ('Lionel Hill', 55, Decimal('89748.4945'), 'older than average'),
 ('Quemby Berg', 47, Decimal('91385.4815'), 'little over average age'),
 ('Byron Randolph', 56, Decimal('89183.4086'), 'older than average'),
 ('Penelope Alexander', 38, Decimal('93187.0984'), 'younger than average'),
 ('Victor Reed', 32, Decimal('89467.0526'), 'younger than average'),
 ('Carla Avery', 55, Decimal('89748.4945'), 'older than average'),
 ('Macaulay Rivers', 39, Decimal('93658.6212'), 'little over average age'),
 ('Olivia Cameron', 35, Decimal('90311.6364'), 'younger than average'),
 ('Aimee Estrada', 44, Decimal('92423.0526'), 'little over average age'),
 ('Eagan Buck', 38, Decimal('93187.0984'), 'younger than 

Discovery: What I discovered about this query is that the youngest employees are generally making more money. Because I ordered the query by salary descending, the highest salaries are first on the list. From this, you can see that the top salaries are the younger than average employees.

# #2
Purpose: This query is to showcase how much potential revenue is on hand for the products available. This could help with financial estimations as well as inventory.

In [22]:
%%sql
SELECT 
	product_name ,
    quantity_on_hand,
    price,
	price * quantity_on_hand AS dollar_amount_available
FROM products p
ORDER BY dollar_amount_available DESC
LIMIT 10; 

 * mysql://admin:***@lmu-dev-01.cniirpmw0saj.us-east-1.rds.amazonaws.com/sql_project
10 rows affected.


[('WD Austerity 2-8-0 90448 BR Black Late Crest', 399, Decimal('135.00'), Decimal('53865.00')),
 ('Hornby 00 Gauge 253mm Weathered Paviland Grange Steam Locomotive Train Model', 390, Decimal('119.50'), Decimal('46605.00')),
 ('Hornby Gauge Western Express Digital Train Set with eLink and TTS Loco Train Set', 144, Decimal('235.58'), Decimal('33923.52')),
 ('Kato (USA) 1761308 F3B Denver & Rio Grande Western', 123, Decimal('273.60'), Decimal('33652.80')),
 ('Bachmann 32-882 Fairburn 2-6-4 Tank 42062 BR Lined Black Late Crest', 299, Decimal('99.95'), Decimal('29885.05')),
 ('Power Trains Auto Loader City', 334, Decimal('76.99'), Decimal('25714.66')),
 ('Piko 59497', 138, Decimal('177.99'), Decimal('24562.62')),
 ('Walthers Proto 920-40695 EMD F7A RI No.676', 128, Decimal('167.10'), Decimal('21388.80')),
 ('Hornby R3246TTS LNER 2-8-2 Cock O The North P2 Class with TTS Sound', 147, Decimal('139.95'), Decimal('20572.65')),
 ('Hornby 00 Gauge 247mm BR Gresley Teak Buffet Coach Model', 294, Decimal('60.25'), Decimal('17713.50'))]

Discovery: From this query, the business can use this information to decide what products should be marketed more or less. For instance, there is $53,865.00 worth of potential sales coming directly from the WD Austerity 2-8-0 90448 BR Black Late Crest product. Sales and Marketing departments should focus on promoting this product, as well as the other top 10 performers.

# #3
Purpose: This query would benefit the business's customers because they would be able to see what location in Los Angeles provides the highest quantity of goods.

In [29]:
%%sql
SELECT 
	sd.store_id, 
	name, 
	city, 
	state ,
	SUM(quantity_on_hand) AS highest_products_quantity_available
FROM store_data sd 
JOIN product_store ps
	ON sd.store_id = ps.store_id 
JOIN products p 
	ON ps.product_id = p.product_id 
GROUP BY sd.store_id
ORDER BY highest_products_quantity_available DESC
;

 * mysql://admin:***@lmu-dev-01.cniirpmw0saj.us-east-1.rds.amazonaws.com/sql_project
16 rows affected.


[(15, 'Costco Pharmacy', 'Inglewood', 'CA', Decimal('5253')),
 (13, 'Costco Gasoline', 'Marina del Rey', 'CA', Decimal('5229')),
 (9, 'Costco Wholesale', 'Alhambra', 'CA', Decimal('4845')),
 (14, 'Costco Gasoline', 'Inglewood', 'CA', Decimal('4546')),
 (8, 'Costco Business Center', 'Hawthorne', 'CA', Decimal('4444')),
 (6, 'Costco', 'Woodland Hills', 'CA', Decimal('4413')),
 (3, 'Costco', 'Los Angeles', 'CA', Decimal('4399')),
 (5, 'Costco Wholesale', 'Inglewood', 'CA', Decimal('4393')),
 (16, 'Costco Business Center - Gasoline', 'Hawthorne', 'CA', Decimal('4281')),
 (12, 'Costco Gasoline', 'Hawthorne', 'CA', Decimal('4267')),
 (7, 'Costco Wholesale', 'Burbank', 'CA', Decimal('3700')),
 (10, 'Costco Wholesale', 'Signal Hill', 'CA', Decimal('3530')),
 (11, 'Costco', 'Van Nuys', 'CA', Decimal('3226')),
 (2, 'Costco Wholesale', 'Hawthorne', 'CA', Decimal('2938')),
 (4, 'Costco Wholesale', 'Torrance', 'CA', Decimal('2808')),
 (1, 'Costco Wholesale', 'Marina Del Rey', 'CA', Decimal('2590'))]

Discovery: I discovered in this query that the highest quantity of goods offered is from a Costco Pharmacy location in Inglewood. This output surprised me and I believe would surprise many people. 

# #4
Purpose: This query is meant to display the highest earning employee and the lowest earning employee. This would be helpful for a business to know because managers can understand the dispersion of salary.

In [38]:
%%sql
SELECT 
    employee_id,
    CONCAT(first_name, ' ', last_name) AS full_name,
    age,
    salary
FROM employees
WHERE salary = (
    SELECT MAX(salary)
    FROM employees
)
    OR salary = (
    SELECT MIN(salary)
    FROM employees
);

 * mysql://admin:***@lmu-dev-01.cniirpmw0saj.us-east-1.rds.amazonaws.com/sql_project
2 rows affected.


[(49, 'Charlie Parker', 30, 192968), (239, 'Samson Sparks', 58, 30722)]

Discovery: I discovered from this query that there is a much higher difference between employee salary than I initially thought. From here, the business can dig deeper into what each of their positions are, educational background, age, as well as other factors that effect compensation.

# #5
Purpose: This query outputs the employee information for those who make higher than average compensation (salary). This is beneficial for the business because we can use these employees as a guide for bottom-tier employees who make less than average. I chose to limit the results by 10, ordering by salary descending, in order to only show the top 10 highest salaried employees.

In [41]:
%%sql
SELECT *
FROM employees e 
WHERE salary > (
	SELECT AVG(salary)
	FROM employees e
)
ORDER BY salary DESC
LIMIT 10;

 * mysql://admin:***@lmu-dev-01.cniirpmw0saj.us-east-1.rds.amazonaws.com/sql_project
10 rows affected.


[(49, 'Charlie', 'Parker', 30, 'c.parker@costco.com', 786501550, 192968, 15),
 (48, 'Ellia', 'Turner', 35, 'e.turner@costco.com', 98610757, 189551, 16),
 (28, 'Maria', 'Taylor', 30, 'm.taylor@costco.com', 209989703, 180375, 9),
 (45, 'Owen', 'Howard', 36, 'o.howard@costco.com', 607917260, 179672, 2),
 (47, 'Jordan', 'Hawkins', 36, 'j.hawkins@costco.com', 702484660, 175303, 3),
 (2, 'Dainton', 'Payne', 37, 'd.payne@costco.com', 189622500, 175179, 1),
 (19, 'Emily', 'Stewart', 31, 'e.stewart@costco.com', 245855658, 175107, 8),
 (37, 'Martin', 'Russell', 37, 'm.russell@costco.com', 13960274, 172697, 11),
 (14, 'Miranda', 'Ellis', 33, 'm.ellis@costco.com', 904029829, 170029, 4),
 (7, 'Natalie', 'Moore', 40, 'n.moore@costco.com', 697280895, 167917, 7)]

Discovery: I found that from this query, the top highest earning employees are within the ages of 30-40. This is quite shocking to me because I would initially think that higher earning employees would have worked at the company for 20+ years. Costco is very impressive when it comes to employee retention because they focus on hiring from within and having great rewards. From this query, Costco can compare these results with how long they've been at the company as well as what position they hold.

# Primary Question Query 
I chose to answer the broad question, "What Costco store in the Los Angeles area is best?" because I believe there are many different factors that go into answering it. Each individual is different and has differing needs. Some people value employee interaction/help, some value product selection, and some base the decision entirely upon other's opinions. With this query, I show all possible values to find try and find a correlation with rating from Yelp. I first approached this query by understanding and plotting what columns were most necessary in order to convey the information. I decided to use COUNT() in order to see how many employees work at each store. Next, I used SUM() to add up the total quantity of products that are being held at each location. I used this number rather than counting the number of products because I decided that total quantity would be more beneficial. I then aliased the calculations to improve readability. From there, I looked at the ERD of the database and used multiple JOINs to connect them. The GROUP BY function was lastly used in the CTE to be sure each aggregate function is being properly calculated. I decided to use a Common Table Expression because I find this to be easier to use and read. After the CTE is formatted, I used the AVG() Windows Function to calculate the average total quantity per rating. This will help understand the differences between locations and ratings. Lastly, I ordered the results first by rating and then by total quantity, both of which descending.

In [ ]:
%%sql 
WITH best_costco_cte AS(
	SELECT 
		sd.store_id ,
		name,
		city,
		rating,
		COUNT(employee_id) AS num_of_employee,
		SUM(quantity_on_hand) AS total_quantity_in_store
	FROM employees e 
	JOIN store_data sd 
		ON e.store_id = sd.store_id
	JOIN product_store ps 
		ON sd.store_id = ps.store_id 
	JOIN products p 
		ON ps.product_id = p.product_id 
	GROUP BY sd.store_id 
)
SELECT 
	*,
	AVG(total_quantity_in_store) OVER(
		PARTITION BY rating
	) AS avg_total_quantity_per_rating
FROM best_costco_cte
ORDER BY 
    rating DESC,
    total_quantity_in_store DESC
;

From the query above, I do not see any direct correlations between rating, total quantity, and number of employees. Going on, I would recommend to conduct interviews with the employees, look at the size of the store itself, or look at what other additions Costco has.

# Related Question #1 Query


What location in Los Angeles has the highest rating and supplies the largest selection of products?

I chose to combine questions relating to rating and total number of products available at each location because I wanted to see if there is a strong correlation between the two. This would help Costco understand if the larger the quantity of goods a store has directly relates to a higer rating from Yelp. I first began this problem by SELECTing the columns I knew would be helpful for labeling/understanding the output and those that would be necessary. I used the aggregate function, COUNT(), to select the number of total products (using product_id) found in each store (using store_id). In order to pull from the `product_store` table, I needed to use a JOIN function, connecting the two by using the `store_id` column. I chose to use a Common Table Expression so that I coul easily alias the table names. This will make it easier for analysts to understand the output. I used the aggregate windows function, AVG(), so that I could partition the data by `rating`. I aliased that output as avg_num_of_products_per_rating. Lastly, I ordered the data first by `rating` and second by num_of_products, both in descending order.

In [51]:
%%sql
WITH product_num_cte AS(
	SELECT 
		sd.store_id ,
		name,
		city, 
		rating,
		COUNT(ps.product_id) AS num_of_products
	FROM store_data sd 
	JOIN product_store ps 
		ON sd.store_id = ps.store_id 
	GROUP BY sd.store_id
)
SELECT 
	name,
	city,
	rating,
    num_of_products,
	AVG(num_of_products) OVER(
		PARTITION BY rating) AS avg_num_of_products_per_rating
FROM product_num_cte
ORDER BY 
    rating DESC,
    num_of_products DESC;

 * mysql://admin:***@lmu-dev-01.cniirpmw0saj.us-east-1.rds.amazonaws.com/sql_project
16 rows affected.


[('Costco Business Center - Gasoline', 'Hawthorne', 5, 40, Decimal('40.0000')),
 ('Costco Gasoline', 'Hawthorne', 4, 47, Decimal('35.0000')),
 ('Costco Gasoline', 'Inglewood', 4, 40, Decimal('35.0000')),
 ('Costco Business Center', 'Hawthorne', 4, 38, Decimal('35.0000')),
 ('Costco Wholesale', 'Torrance', 4, 32, Decimal('35.0000')),
 ('Costco Wholesale', 'Hawthorne', 4, 28, Decimal('35.0000')),
 ('Costco Wholesale', 'Marina Del Rey', 4, 25, Decimal('35.0000')),
 ('Costco Gasoline', 'Marina del Rey', 3, 47, Decimal('36.4444')),
 ('Costco Pharmacy', 'Inglewood', 3, 40, Decimal('36.4444')),
 ('Costco', 'Woodland Hills', 3, 39, Decimal('36.4444')),
 ('Costco Wholesale', 'Alhambra', 3, 38, Decimal('36.4444')),
 ('Costco Wholesale', 'Signal Hill', 3, 38, Decimal('36.4444')),
 ('Costco Wholesale', 'Inglewood', 3, 35, Decimal('36.4444')),
 ('Costco', 'Los Angeles', 3, 32, Decimal('36.4444')),
 ('Costco Wholesale', 'Burbank', 3, 31, Decimal('36.4444')),
 ('Costco', 'Van Nuys', 3, 28, Decimal('36.4444'))]

From this related query, I based the overarching question off of both rating and number of products sold. I now recommend that Costco uses the Costco Business Center- Gasoline location as a business model to improve the rest. From the query, I can see that the top 3 locations have high ratings as well as high number of products available. I cannot tell, however, that there is a strong correlation between number of products available and rating. The second location 'Costco Gasoline' in Hawthorn has a 4 star rating and provides a larger array of products, far surpassing the average number per 4 star rating. As you can also see, the average number of products per rating is higher for 3 stars than 4 star ratings. Overall, I would recommend to look at other factors from Hawthron and Inglewood that could provide more insight into higher rating.

# Related Question #2 Query
What location has the most employees working per rating?

I chose to look at the number of employees working at each location per rating in order to see if there is a correlation between employees working and customer satisfaction (rating). This would help Costco deduce if they should hire more employees to be of service to customers. This would potentially lead to higher ratings. I chose to use a VIEW because I wanted to easily save the table for other people to easily use. This would be beneficial to analysts because it saves time and prevents unnecessary mistakes. I began the query by mapping out how I was going to recieve the columns, looking at each table present. I used a JOIN function in order to pull the store data as well as `employee_id`. I then used the COUNT() aggregate function in order to calculate the number of employees that work at each location. Lastly, I grouped by `store_id` to be sure there is no repatition.

Next, after creating the VIEW, I wanted to confirm that the VIEW was working. I knew I was going to use the AVG() function in a later query, so I decided to find the average number of employees working throughout all locations.

Finally, I decided to use the AVG() aggregate function so I can put the total number of employees working into perspective. I then used the CASE function to label each location depending on the average number of employees per rating. I used the PARTITION BY function to make sure it was by rating. Lastly, I used the ORDER BY function to make sure the rating is the initial order, then within the rating ordering by the number of employees working. Both rating and number of employees ordered descending.

In [57]:
%%sql
CREATE OR REPLACE VIEW employee_store_data AS
	SELECT 
		sd.store_id, 
		name,
		city,
		zip_code,
		rating,
		COUNT(employee_id) AS num_of_employees_working
	FROM store_data sd 
	JOIN employees e
		ON sd.store_id = e.store_id 
	GROUP BY sd.store_id ;

 * mysql://admin:***@lmu-dev-01.cniirpmw0saj.us-east-1.rds.amazonaws.com/sql_project
0 rows affected.


[]

In [60]:
%%sql
SELECT AVG(num_of_employees_working) AS avg_num_of_employees_working
FROM employee_store_data;

 * mysql://admin:***@lmu-dev-01.cniirpmw0saj.us-east-1.rds.amazonaws.com/sql_project
1 rows affected.


[(Decimal('6.0000'),)]

In [69]:
%%sql
SELECT 
	store_id,
	name,
	rating,
	num_of_employees_working,
    AVG(num_of_employees_working) OVER(
        PARTITION BY rating
    ) AS avg_employees_per_rating,
	CASE 
		WHEN num_of_employees_working < (
			SELECT AVG(num_of_employees_working)
			FROM employee_store_data
		) THEN 'less_than_average_employees'
		WHEN num_of_employees_working > (
			SELECT AVG(num_of_employees_working)
			FROM employee_store_data
		) THEN 'more_than_average_employees'
		ELSE 'average_num_of_employees'
	END AS num_of_employees_label
FROM employee_store_data
ORDER BY 
	rating DESC,
	num_of_employees_working DESC;

 * mysql://admin:***@lmu-dev-01.cniirpmw0saj.us-east-1.rds.amazonaws.com/sql_project
16 rows affected.


[(16, 'Costco Business Center - Gasoline', 5, 6, Decimal('6.0000'), 'average_num_of_employees'),
 (4, 'Costco Wholesale', 4, 10, Decimal('5.5000'), 'more_than_average_employees'),
 (1, 'Costco Wholesale', 4, 8, Decimal('5.5000'), 'more_than_average_employees'),
 (2, 'Costco Wholesale', 4, 5, Decimal('5.5000'), 'less_than_average_employees'),
 (14, 'Costco Gasoline', 4, 5, Decimal('5.5000'), 'less_than_average_employees'),
 (8, 'Costco Business Center', 4, 3, Decimal('5.5000'), 'less_than_average_employees'),
 (12, 'Costco Gasoline', 4, 2, Decimal('5.5000'), 'less_than_average_employees'),
 (7, 'Costco Wholesale', 3, 10, Decimal('6.3333'), 'more_than_average_employees'),
 (11, 'Costco', 3, 10, Decimal('6.3333'), 'more_than_average_employees'),
 (3, 'Costco', 3, 9, Decimal('6.3333'), 'more_than_average_employees'),
 (6, 'Costco', 3, 7, Decimal('6.3333'), 'more_than_average_employees'),
 (10, 'Costco Wholesale', 3, 5, Decimal('6.3333'), 'less_than_average_employees'),
 (13, 'Costco Gasoline', 3, 5, Decimal('6.3333'), 'less_than_average_employees'),
 (9, 'Costco Wholesale', 3, 4, Decimal('6.3333'), 'less_than_average_employees'),
 (15, 'Costco Pharmacy', 3, 4, Decimal('6.3333'), 'less_than_average_employees'),
 (5, 'Costco Wholesale', 3, 3, Decimal('6.3333'), 'less_than_average_employees')]

From the query, I can see once more that there is no serious correlation between number of employees and rating. The average number of employees per rating is surprisingly higher. From this information, I would recommend that Costco looks further into other factors that could effect rating. 